### PR TITLE
Bring code more up to current

### PR DESCRIPTION
**Pull request by marcelvriend**
## Proposed changes

- Use entity descriptions for sensor and binary sensor platform.
- Replace the custom JSON encoder with the HA default.
- Remove the broken mechanism to retrieve Grocy data only when it has been modified.
- Remove unused translation strings. The description with URL is redundant with the help icon and better not to have URLs in language specific files.
- Set missing unit of measurement for shopping list and stock sensor.
- Hide entities in HA for disabled features in Grocy.
- Request for enabled Grocy features only on integration initialization.
- Rewrite the entity refresh in some service calls to be not entity id specific.
- Split the coordinator class into a separate file.
- Clean up code.

No breaking changes.

## Out of scope issues

- With all entities enabled and having a reasonable amount of data, there are a lot of API requests every 30 seconds. Already improved it a bit, but it is still a lot.

The mechanism to prevent this was already broken, so nothing has changed in that regard. Although easy to fix the bug, it still ignores time zones and will likely cause refresh issues for many users. Needs more attention and therefore removed it for now.

### CODE REVIEW

**Comment by isabellaalstrom** [comment]
If we can do it like @ludeeus says, without breaking change that would be best. Otherwise like you say, a production release of current beta before. I might do that now anyway? I am running beta with no issues.

File: custom_components/grocy/__init__.py [hunk 1]
```diff
@@ -4,151 +4,104 @@
...
-        config_entry.data[CONF_PORT],
-        config_entry.data[CONF_VERIFY_SSL],
+    coordinator: GrocyDataUpdateCoordinator = GrocyDataUpdateCoordinator(hass)
+    coordinator.available_entities = await _async_get_available_entities(
+        coordinator.grocy_data
     )
-
     await coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = coordinator
```
Comment (ludeeus) [inline custom_components/grocy/__init__.py:47]: This nested dictionary is only needed if the integration can be configured multiple times, which is not currently supported by this integration.

File: custom_components/grocy/entity.py [hunk 2]
```diff
@@ -1,108 +1,44 @@
...
-        elif self.entity_type == GrocyEntityType.OVERDUE_TASKS:
-            data = {"tasks": [x.as_dict() for x in self.entity_data]}
-        elif self.entity_type == GrocyEntityType.PRODUCTS:
-            data = {"products": [x.as_dict() for x in self.entity_data]}
-        elif self.entity_type == GrocyEntityType.SHOPPING_LIST:
-            data = {"products": [x.as_dict() for x in self.entity_data]}
-        elif self.entity_type == GrocyEntityType.STOCK:
-            data = {"products": [x.as_dict() for x in self.entity_data]}
-        elif self.entity_type == GrocyEntityType.TASKS:
-            data = {"tasks": [x.as_dict() for x in self.entity_data]}
```
Comment (ludeeus) [inline custom_components/grocy/entity.py:103]: Awesome! 🙈

File: custom_components/grocy/entity.py [hunk 3]
```diff
@@ -1,108 +1,44 @@
...
-        return GrocyEntityIcon.DEFAULT
+    def device_info(self) -> DeviceInfo:
+        """Grocy device information."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.coordinator.config_entry.entry_id)},
+            name=NAME,
+            manufacturer=NAME,
+            model=VERSION,  # TODO This should be the Grocy version
+            entry_type=DeviceEntryType.SERVICE,
+            configuration_url=self.coordinator.config_entry.data[CONF_URL],
```
Comment (ludeeus) [inline custom_components/grocy/entity.py:29]: FYI this does not work if configured with the internal hostname for the add-on as the URL

File: custom_components/grocy/entity.py [hunk 4]
```diff
@@ -1,108 +1,44 @@
...
-            return GrocyEntityIcon[GrocyEntityType(self.entity_type).name]
-
-        return GrocyEntityIcon.DEFAULT
+    def device_info(self) -> DeviceInfo:
+        """Grocy device information."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.coordinator.config_entry.entry_id)},
+            name=NAME,
+            manufacturer=NAME,
+            model=VERSION,  # TODO This should be the Grocy version
```
Comment (ludeeus) [inline custom_components/grocy/entity.py:27]: `model` should probably not be used at all?
There are a `software_version` key that fits better.

File: custom_components/grocy/binary_sensor.py [hunk 5]
```diff
@@ -1,45 +1,136 @@
...
+        self,
+        coordinator: GrocyDataUpdateCoordinator,
+        description: BinarySensorEntityDescription,
+        config_entry: ConfigEntry,
+    ) -> None:
+        """Initialize binary sensor entity."""
+        super().__init__(coordinator)
+        self._attr_name = description.name
+        self._attr_unique_id = f"{config_entry.entry_id}{description.key.lower()}"
+        self.entity_description = description
```
Comment (ludeeus) [inline custom_components/grocy/binary_sensor.py:129]: This is the same for the sensor as well (other than typing for the description), so this should move to `GrocyEntity`

**Comment by ludeeus** [comment]
As I do not use this integration, and further reviews have to be done by @isabellaalstrom

**Comment by isabellaalstrom** [comment]
Oh crap I missed this entirely! On vacay right now. Will look at it asap.

**Comment by isabellaalstrom** [comment]
I looked the code over but I'm not all that great with python really (😬) so I'm gonna merge it, release a pre-release and try it out!

**Comment by isabellaalstrom** [comment]
This should be version 5.0.0, right? Since breaking etc

**Pull request merged**